Improving Performance and creation of intances

ChicK00o opened the pull request:
Have a look and let me know

ChicK00o committed:
Merge pull request #1 from Disturbing/master

Pulling in the new changes from Coop to suice

RohitBhosle-Housing committed:
Improving the CreateInstance call by changing the implementation. Further improving performance

disturbing commented on DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs:
```diff
@@ -0,0 +1,70 @@
+    {
+        Type[] argTypes = args.Select(a => a.GetType()).ToArray();
+
+        var ctor = type.GetConstructor(argTypes);
```
what is ctro - try to use full names and define variables versus using var.

ChicK00o replied:
My bad :)

ChicK00o committed:
Removing "vars" and using descriptive names

ChicK00o commented:
Have a look again

disturbing commented on DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs:
```diff
@@ -0,0 +1,70 @@
+
+    internal delegate object ObjectActivator(params object[] args);
+
+    private static Dictionary<ConstructorInfo, object> _activators = new Dictionary<ConstructorInfo, object>();
```
Move dictionary to top of class and delegate definition as per code format rules for this project.

disturbing replied:
remove _ as private variable - this project doesn't do that in terms of format.

disturbing commented on DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs:
```diff
@@ -0,0 +1,70 @@
+    internal static ObjectActivator GetActivator(ConstructorInfo ctor)
+    {
+        object activator;
+        if (_activators.TryGetValue(ctor, out activator)) return (ObjectActivator)activator;
```
No need to store in dictionary, object - you can store in dictionary, ObejctActivator, so we don't need to have a cast here.
- Also for this project I've tried to have single return out per function rule Please reformat to do the following:

```
ObjectActivator activator;

if (TryGetValue == false){
     activator = CreateNewActivator(params);
}

return activator;


```

disturbing commented on DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs:
```diff
@@ -0,0 +1,70 @@
+        //pick each arg from the params array
+        //and create a typed expression of them
+        for (int i = 0; i < paramsInfo.Length; i++)
+        {
```
Same line brackets.

disturbing commented on DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs:
```diff
@@ -0,0 +1,70 @@
+        ParameterInfo[] paramsInfo = ctor.GetParameters();
+
+        //create a single param of type object[]
+        ParameterExpression param = Expression.Parameter(typeof(object[]), "args");
```
use constant for name 'ARGS'.

disturbing commented on DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs:
```diff
@@ -0,0 +1,70 @@
+        LambdaExpression lambda = Expression.Lambda(typeof(ObjectActivator), newExp, param);
+
+        //compile it
+        var compiled = (ObjectActivator)lambda.Compile();
```
static defined types, remove var please.

disturbing commented on DTools/Suice/suice/Utilities/InstanceCreatorExtension.cs:
```diff
@@ -0,0 +1,70 @@
+/// https://rboeije.wordpress.com/2012/04/19/extension-method-on-type-as-alternative-for-activator-createinstance/ - Implemented this
+/// 
+/// This extension is a benchmarked better implementation to Activator.CreateInstance when taking the second link into consideration
+/// Here the expression is further cached so as to better instantiation for factory implementations
```
Add your github name as @author

Disturbing commented:
Sorry, I was not able to finish the full review until now.

I just wrote the remaining comments.  Once all comments are resolved, I will accept the pull request.

Great work overall.

ChicK00o committed:
Dictornary type changed to remove casting.
changes to format and code arrangement, to be in sync with the rest of Coop's code style

ChicK00o commented:
Incorporated the changes you mentioned, have a look again and
Let me know if there are any further changes you need
Cheers

Disturbing commented:
All good :+1:
